File: dotnet/test/Azure.Iot.Operations.Protocol.MetlTests/TestCasePrologue.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Azure.Iot.Operations.Protocol.MetlTests
{
    public class TestCasePrologue
    {
        public TestCaseMqttConfig? MqttConfig { get; set; }

        public TestCasePushAcks? PushAcks { get; set; }

        public List<TestCaseExecutor> Executors { get; set; } = new();

        public List<TestCaseInvoker> Invokers { get; set; } = new();

        public List<TestCaseReceiver> Receivers { get; set; } = new();

        public List<TestCaseSender> Senders { get; set; } = new();

        public TestCaseCatch? Catch { get; set; }

        public Dictionary<string, int> CountdownEvents { get; set; } = new();
    }
}

File: go/test/protocol/test_case_prologue.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package protocol

import (
	"gopkg.in/yaml.v3"
)

type testCasePrologue struct {
	MqttConfig      TestCaseMqttConfig `yaml:"mqtt-config"`
	PushAcks        TestCasePushAcks   `yaml:"push-acks"`
	Executors       []TestCaseExecutor `yaml:"executors"`
	Invokers        []TestCaseInvoker  `yaml:"invokers"`
	Receivers       []TestCaseReceiver `yaml:"receivers"`
	Senders         []TestCaseSender   `yaml:"senders"`
	Catch           *TestCaseCatch     `yaml:"catch"`
	CountdownEvents map[string]int     `yaml:"countdown-events"`
}

type TestCasePrologue struct {
	testCasePrologue
}

func (prologue *TestCasePrologue) UnmarshalYAML(node *yaml.Node) error {
	*prologue = TestCasePrologue{}

	prologue.MqttConfig = MakeTestCaseMqttConfig()

	return node.Decode(&prologue.testCasePrologue)
}
